README
Mojolicious::Plugin::MySQLViewerLite

  Mojolicious plugin to display mysql database information on browser

SYNOPSYS

  # Mojolicious::Lite
  plugin 'MySQLViewerLite', dbh => $dbh;

  # Mojolicious
  $app->plugin('MySQLViewerLite', dbh => $dbh);

  # Access
  http://localhost:3000/mysqlviewerlite
  
DESCRIPTION

  Mojolicious::Plugin::MySQLViewerLite is Mojolicious plugin
  to display MySQL database information on browser.

  This module have the following features.

    - Display all table names
    - Display show create table
    - Select * from TABLE
    - Display primary keys, null allowed columnes,
      database engines, and charsets in all tables.


Changes
0.15 (DEPRECATED)
  - this module is merged into Mojolicious::Plugin::DBViewer.
  - this module is DEPRECATED and will be removed from CPAN in 2018/4/1.
0.14
  - removed Mojolicious waypoint logic because waypoint is deprecated.
0.13
  - catch up DBIx::Custom 0.25 version up
0.12
  - improved show create table page design.
  - improved select page design. 
0.11
  - improved design
  - added show select tables page
0.10
  - added dbi option
0.09
  - fixed test
0.08
  - added simple table search
  - added connector option
  - added page navigation
0.07
  - cleanup modules
0.06
  - added show create tables page
0.05
  - fixed bug that 'select * from TABLE' is not displayed
    if database is not current database
0.04
  - improved design, removed primary key last commma.
  - display charsets in all talbes.
  - suppress warnings
0.03
  - move README.pod to README.
0.02
  - added more tests.
0.01
  - First release.


t/test.pl
use Mojolicious::Lite;

use DBIx::Custom;
use FindBin;
use lib "lib";

use Mojolicious::Plugin::MySQLViewerLite;
my $dbi = DBIx::Custom->connect(
  dsn => 'dbi:mysql:database=dbix_custom',
  user => 'dbix_custom',
  password => 'dbix_custom'
);

plugin 'MySQLViewerLite', dbh => $dbi->dbh, prefix => 'mysqlviewer';

get '/' => {text => 'a'};

app->start;



Makefile.PL
use 5.006;
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME                => 'Mojolicious::Plugin::MySQLViewerLite',
    AUTHOR              => q{Yuki Kimoto <kimoto.yuki@gmail.com>},
    VERSION_FROM        => 'lib/Mojolicious/Plugin/MySQLViewerLite.pm',
    ABSTRACT_FROM       => 'lib/Mojolicious/Plugin/MySQLViewerLite.pm',
    ($ExtUtils::MakeMaker::VERSION >= 6.3002
      ? ('LICENSE'=> 'perl')
      : ()),
    PL_FILES            => {},
    PREREQ_PM => {
        'Test::More' => 0,
        'Mojolicious' => '2.59',
        'Validator::Custom' => '0.1426',
        'DBIx::Custom' => '0.25',
        'Data::Page' => '2.02'
    },
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean               => { FILES => 'Mojolicious-Plugin-MySQLViewerLite-*' },
);


lib/Mojolicious/Plugin/MysqlViewerCheap.pm
package Mojolicious::Plugin::MysqlViewerCheap;
use Mojo::Base 'Mojolicious::Plugin';

use DBIx::Custom;
use Validator::Custom;

# Validator
my $vc = Validator::Custom->new;
$vc->register_constraint(
    safety_name => sub {
        my $name = shift;
        
        return ($name || '') =~ /^\w+$/ ? 1 : 0;
    }
);

# DBI 
my $dbi;

my %args = (template_class => __PACKAGE__);
sub register {
    my ($self, $app, $conf) = @_;
    
    my $dbh = $conf->{dbh};
    my $r = $conf->{route} || $app->routes;
    
    $dbi = DBIx::Custom->new;
    $dbi->dbh($dbh);
    
    # Top page
    $r->get('/mysqlviewer', sub {
        my $self = shift;
        my $stash = $self->stash;
        
        $stash->{databases} = _show_databases();
        $stash->{current_database} = _current_database();
        
        return $self->render(%args);
    });
    
    # Database
    $r->get('/mysqlviewer/database', sub {
        my $self = shift;
        
        my $param = $self->req->params->to_hash;
        my $rule = [
            database => {default => ''} => [
                'safety_name'
            ] 
        ];
        my $vresult = $vc->validate($param, $rule);
        my $database = $vresult->data->{database};
        
        my $tables = _show_tables($database);
        
        return $self->render(%args, database => $database, tables => $tables);
    } => 'mysqlviewer-database');
    
    # Table
    $r->get('/mysqlviewer/table', sub {
        my $self = shift;
        
        # Validation
        my $param = $self->req->params->to_hash;
        my $rule = [
            database => {default => ''} => [
                'safety_name'
            ],
            table => {default => ''} => [
                'safety_name'
            ]
        ];
        my $vresult = $vc->validate($param, $rule);
        my $database = $vresult->data->{database};
        my $table = $vresult->data->{table};
        
        my $table_def = _show_create_table($database, $table);
        return $self->render(%args, database => $database, table => $table, 
          table_def => $table_def, current_database => _current_database());
    } => 'mysqlviewer-table');
    
    # List primary keys
    $r->get('/mysqlviewer/listprimarykeys', sub {
        my $self = shift;
        
        # Validation
        my $param = $self->req->params->to_hash;
        my $rule = [
            database => {default => ''} => [
                'safety_name'
            ],
        ];
        my $vresult = $vc->validate($param, $rule);
        my $database = $vresult->data->{database};
        
        # Get primary keys
        my $tables = _show_tables($database);
        my $primary_keys = {};
        foreach my $table (@$tables) {
            my $show_create_table = _show_create_table($database, $table) || '';
            my $primary_key = '';
            if ($show_create_table =~ /PRIMARY\s+KEY\s+(.+?)\n/i) {
                $primary_key = $1;
            }
            $primary_keys->{$table} = $primary_key;
        }
        
        $self->render(%args, database => $database, primary_keys => $primary_keys);
        
    } => 'mysqlviewer-listprimarykeys');

    # List null allowed columns
    $r->get('/mysqlviewer/listnullallowedcolumns', sub {
        my $self = shift;
        
        # Validation
        my $param = $self->req->params->to_hash;
        my $rule = [
            database => {default => ''} => [
                'safety_name'
            ],
        ];
        my $vresult = $vc->validate($param, $rule);
        my $database = $vresult->data->{database};
        
        # Get null allowed columns
        my $tables = _show_tables($database);
        my $null_allowed_columns = {};
        foreach my $table (@$tables) {
            my $show_create_table = _show_create_table($database, $table) || '';
            my @lines = split(/\n/, $show_create_table);
            my $null_allowed_column = [];
            foreach my $line (@lines) {
                next if /^\s*`/ || $line =~ /NOT\s+NULL/i;
                if ($line =~ /^\s+(`\w+?`)/) {
                    push @$null_allowed_column, $1;
                }
            }
            $null_allowed_columns->{$table} = $null_allowed_column;
        }
        
        $self->render(%args, database => $database,
          null_allowed_columns => $null_allowed_columns);
        
    } => 'mysqlviewer-listnullallowedcolumns');

    # List database engines
    $r->get('/mysqlviewer/listdatabaseengines', sub {
        my $self = shift;
        
        # Validation
        my $param = $self->req->params->to_hash;
        my $rule = [
            database => {default => ''} => [
                'safety_name'
            ],
        ];
        my $vresult = $vc->validate($param, $rule);
        my $database = $vresult->data->{database};
        
        # Get null allowed columns
        my $tables = _show_tables($database);
        my $database_engines = {};
        foreach my $table (@$tables) {
            my $show_create_table = _show_create_table($database, $table) || '';
            my $database_engine = '';
            if ($show_create_table =~ /ENGINE=(.+?)\s+/i) {
                $database_engine = $1;
            }
            $database_engines->{$table} = $database_engine;
        }
        
        $self->render(%args, database => $database,
          database_engines => $database_engines);
        
    } => 'mysqlviewer-listdatabaseengines');

    # List database engines
    $r->get('/mysqlviewer/selecttop1000', sub {
        my $self = shift;
        
        # Validation
        my $param = $self->req->params->to_hash;
        my $rule = [
            database => {default => ''} => [
                'safety_name'
            ],
            table => {default => ''} => [
                'safety_name'
            ]
        ];
        my $vresult = $vc->validate($param, $rule);
        my $database = $vresult->data->{database};
        my $table = $vresult->data->{table};
        
        # Get null allowed columns
        my $result = $dbi->select(table => $table, append => 'limit 0, 1000');
        my $header = $result->header;
        my $rows = $result->fetch_all;
        my $sql = $dbi->last_sql;
        
        $self->render(%args, database => $database, table => $table,
          header => $header, rows => $rows, sql => $sql);
    } => 'mysqlviewer-selecttop1000'); 
}

sub _current_database {
    $dbi->execute('select database()')->fetch->[0];
} 

sub _show_databases {
    
    my $databases = [];
    my $database_rows = $dbi->execute('show databases')->all;
    foreach my $database_row (@$database_rows) {
        push @$databases, $database_row->{(keys %$database_row)[0]};
    }
    return $databases; 
}

sub _show_tables { 
    my $database = shift;
    my $table_rows;
    eval { $table_rows = $dbi->execute("show tables from $database")->all };
    $table_rows ||= [];
    my $tables = [];
    foreach my $table_row (@$table_rows) {
        push @$tables, $table_row->{(keys %$table_row)[0]};
    }
    return $tables;
}

sub _show_create_table {
    my ($database, $table) = @_;
    my $table_def_row;
    eval { $table_def_row = $dbi->execute("show create table $database.$table")->one };
    $table_def_row ||= {};
    my $table_def = $table_def_row->{'Create Table'} || '';
    return $table_def;
}


1;

__DATA__

@@ layouts/mysqlviewer.html.ep
<!doctype html><html>
  <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" >
  <head>
    <title><%= $title %></title>
    %= javascript '/js/jquery.js'
    
  </head>
  <body>
    %= content;
  </body>
</html>

@@ mysqlviewer.html.ep
% layout 'mysqlviewer', title => 'MySQL Viewer';
<h2>MySQL Viewer</h2>

<h3>Databases</h3>
<ul>
% foreach my $database (sort @$databases) {
  <li>
    <a href="<%= url_for('/mysqlviewer/database')->query(database => $database) %>"><%= $database %>
    %= $current_database eq $database ? '(current)' : ''
  </li>
% }
</ul>

@@ mysqlviewer-database.html.ep
% layout 'mysqlviewer', title => "Database $database";
<h2>Database <%= $database %>
<h3>Tables</h3>
<ul>
% foreach my $table (sort @$tables) {
  <li><a href="<%= url_for('/mysqlviewer/table')->query(database => $database, table => $table) %>"><%= $table %></a></li>
% }
</ul>

<h3>Utility</h3>
<ul>
  <li><a href="<%= url_for('/mysqlviewer/listprimarykeys')->query(database => $database) %>">List primary keys</a></li>
  <li><a href="<%= url_for('/mysqlviewer/listnullallowedcolumns')->query(database => $database) %>">List null allowed columns</a></li>
  <li><a href="<%= url_for('/mysqlviewer/listdatabaseengines')->query(database => $database) %>">List database engines</a></li>
</ul>

@@ mysqlviewer-table.html.ep
% layout 'mysqlviewer', title => "Table $database.$table";
<h2>Table <%= "$database.$table" %>
<h3>show create table</h3>
<pre><%= $table_def %></pre>

<h3>Utilities</h3>
<ul>
  % if ($database eq $current_database) {
    <li><a href="<%= url_for('/mysqlviewer/selecttop1000')->query(database => $database, table => $table) %>">Select top 1000</a></li>
  % }
</ul>

@@ mysqlviewer-listprimarykeys.html.ep
% layout 'mysqlviewer', title => "$database primary keys";
<h3><%= "$database primary keys" %></h3>
<ul>
  % foreach my $table (sort keys %$primary_keys) {
    <li><a href="<%= url_for('/mysqlviewer/table')->query(database => $database, table => $table) %>"><%= $table %></a> <%= $primary_keys->{$table} %></li>
  % }
</ul>

@@ mysqlviewer-listnullallowedcolumns.html.ep
% layout 'mysqlviewer', title => "$database null allowed columns";
<h3><%= "$database null allowed columns" %></h3>
<ul>
  % foreach my $table (sort keys %$null_allowed_columns) {
    <li><a href="<%= url_for('/mysqlviewer/table')->query(database => $database, table => $table) %>"><%= $table %></a> (<%= join(',', @{$null_allowed_columns->{$table}}) %>)</li>
  % }
</ul>

@@ mysqlviewer-listdatabaseengines.html.ep
% layout 'mysqlviewer', title => "$database database engines";
<h3><%= "$database database engines" %></h3>
<ul>
  % foreach my $table (sort keys %$database_engines) {
    <li><a href="<%= url_for('/mysqlviewer/table')->query(database => $database, table => $table) %>"><%= $table %></a> (<%= $database_engines->{$table} %>)</li>
  % }
</ul>

@@ mysqlviewer-selecttop1000.html.ep
% layout 'mysqlviewer', title => "<%= $table %>: Select top 1000";

<h2>Select top 1000</h2>

<table border="1" cellspacing="0" >
  <tr><td>Table name</td><td><%= $table %></td></tr>
  <tr><td>SQL</td><td><%= $sql %></td></tr>
</table>

<br>

<table border="1" cellspacing="0" >
  <tr>
    % foreach my $h (@$header) {
        <th><%= $h %></th>
    % }
  </tr>
  % foreach my $row (@$rows) {
    <tr>
      % foreach my $data (@$row) {
        <td><%= $data %></td>
      % }
    </tr>
  % }
</table>


lib/Mojolicious/Plugin/MySQLViewerLite/Base.pm
use 5.001001;
package Mojolicious::Plugin::MySQLViewerLite::Base;
use Mojo::Base 'Mojolicious::Plugin';
use DBIx::Custom;
use Validator::Custom;
use File::Basename 'dirname';
use Cwd 'abs_path';
use Carp 'croak';

has 'prefix';
has validator => sub {
  my $validator = Validator::Custom->new;
  $validator->register_constraint(
    safety_name => sub {
      my $name = shift;
      return ($name || '') =~ /^\w+$/ ? 1 : 0;
    }
  );
  return $validator;
};

has dbi => sub { DBIx::Custom->new };

has command => sub { croak "Unimplemented" };

sub register { croak "Unimplemented" }

sub add_template_path {
  my ($self, $renderer, $class) = @_;
  $class =~ s/::/\//g;
  $class .= '.pm';
  my $public = abs_path $INC{$class};
  $public =~ s/\.pm$//;
  push @{$renderer->paths}, "$public/templates";
}

1;


lib/Mojolicious/Plugin/MySQLViewerLite/Command.pm
package Mojolicious::Plugin::MySQLViewerLite::Command;
use Mojo::Base 'Mojolicious::Plugin::MySQLViewerLite::Base::Command';

sub current_database {
  my $self = shift;
  return $self->dbi->execute('select database()')->fetch->[0];
}

sub show_primary_key {
  my ($self, $database, $table) = @_;
  my $show_create_table = $self->show_create_table($database, $table) || '';
  my $primary_key = '';
  if ($show_create_table =~ /PRIMARY\s+KEY\s+(.+?)\n/i) {
    $primary_key = $1;
    $primary_key =~ s/,$//;
  }
  return $primary_key;
}

sub show_null_allowed_column {
  my ($self, $database, $table) = @_;
  
  my $show_create_table = $self->show_create_table($database, $table) || '';
  my @lines = split(/\n/, $show_create_table);
  my $null_allowed_column = [];
  for my $line (@lines) {
    next if /^\s*`/ || $line =~ /NOT\s+NULL/i;
    if ($line =~ /^\s+(`\w+?`)/) {
      push @$null_allowed_column, $1;
    }
  }
  return $null_allowed_column;
}

sub show_database_engine {
  my ($self, $database, $table) = @_;
  
  my $show_create_table = $self->show_create_table($database, $table) || '';
  my $database_engine = '';
  if ($show_create_table =~ /ENGINE=(.+?)(\s+|$)/i) {
    $database_engine = $1;
  }
  
  return $database_engine;
}

sub show_charsets {
  my ($self, $database) = @_;
  
  my $tables = $self->show_tables($database);
  my $charsets = {};
  for my $table (@$tables) {
    my $charset = $self->show_charset($database, $table);
    $charsets->{$table} = $charset;
  }
  
  return $charsets;
}

sub show_charset {
  my ($self, $database, $table) = @_;
  
  my $show_create_table = $self->show_create_table($database, $table) || '';
  my $charset = '';
  if ($show_create_table =~ /CHARSET=(.+?)(\s+|$)/i) {
    $charset = $1;
  }
  
  return $charset;
}

sub show_databases {
  my $self = shift;
  
  my $databases = [];
  my $database_rows = $self->dbi->execute('show databases')->all;
  for my $database_row (@$database_rows) {
    push @$databases, $database_row->{(keys %$database_row)[0]};
  }
  return $databases;
}

sub show_tables { 
  my ($self, $database) = @_;
  my $table_rows;
  eval { $table_rows = $self->dbi->execute("show tables from $database")->all };
  $table_rows ||= [];
  my $tables = [];
  for my $table_row (@$table_rows) {
    push @$tables, $table_row->{(keys %$table_row)[0]};
  }
  return $tables;
}

sub show_create_table {
  my ($self, $database, $table) = @_;
  my $table_def_row;
  eval { $table_def_row = $self->dbi->execute("show create table $database.$table")->one };
  $table_def_row ||= {};
  my $table_def = $table_def_row->{'Create Table'} || '';
  return $table_def;
}

1;


lib/Mojolicious/Plugin/MySQLViewerLite/Controller.pm
package Mojolicious::Plugin::MySQLViewerLite::Controller;
use Mojo::Base 'Mojolicious::Plugin::SQLiteViewerLite::Controller';

sub tables {
  my $self = shift;
  $self->stash->{template} = 'mysqlviewerlite/tables';
  $self->SUPER::tables;
}

sub showdatabaseengines {
  my $self = shift;;
  
  my $plugin = $self->stash->{plugin};
  my $command = $plugin->command;

  # Validation
  my $params = $command->params($self);
  my $rule = [
    database => {default => ''} => [
      'safety_name'
    ],
  ];
  my $vresult = $plugin->validator->validate($params, $rule);
  my $database = $vresult->data->{database};
  
  # Get primary keys
  my $database_engines = $command->show_database_engines($database);
  
  $self->stash->{template} = 'mysqlviewerlite/showdatabaseengines'
    unless $self->stash->{template};

  $self->render(
    database => $database,
    database_engines => $database_engines
  );
}

sub showcharsets {
  my $self = shift;;
  
  my $plugin = $self->stash->{plugin};
  my $command = $plugin->command;

  # Validation
  my $params = $command->params($self);
  my $rule = [
    database => {default => ''} => [
      'safety_name'
    ],
  ];
  my $vresult = $plugin->validator->validate($params, $rule);
  my $database = $vresult->data->{database};
  
  # Get primary keys
  my $charsets = $command->show_charsets($database);
  
  $self->stash->{template} = 'mysqlviewerlite/showcharsets'
    unless $self->stash->{template};

  $self->render(
    database => $database,
    charsets => $charsets
  );
}

1;


lib/Mojolicious/Plugin/MySQLViewerLite/Base/Mysqlviewerlite.pm
package Mojolicious::Plugin::MySQLViewerLite::Base::Mysqlviewerlite;
use Mojo::Base 'Mojolicious::Controller';
use Data::Page;

sub default {
  my $self = shift;
  
  my $plugin = $self->stash->{plugin};
  my $command = $plugin->command;
  
  my $database = $command->show_databases;
  my $current_database = $command->current_database;

  $self->render(
    databases => $database,
    current_database => $current_database,
  );
}

sub tables {
  my $self = shift;;
  
  my $plugin = $self->stash->{plugin};
  my $command = $plugin->command;

  my $params = $command->params($self);
  my $rule = [
    database => {default => ''} => [
      'safety_name'
    ] 
  ];
  my $vresult = $plugin->validator->validate($params, $rule);
  my $database = $vresult->data->{database};
  my $tables = $command->show_tables($database);
  
  return $self->render(
    database => $database,
    tables => $tables
  );
}

sub table {
  my $self = shift;;
  
  my $plugin = $self->stash->{plugin};
  my $command = $plugin->command;

  # Validation
  my $params = $command->params($self);
  my $rule = [
    database => {default => ''} => [
      'safety_name'
    ],
    table => {default => ''} => [
      'safety_name'
    ]
  ];
  my $vresult = $plugin->validator->validate($params, $rule);
  my $database = $vresult->data->{database};
  my $table = $vresult->data->{table};
  
  my $table_def = $command->show_create_table($database, $table);

  return $self->render(
    database => $database,
    table => $table, 
    table_def => $table_def,
  );
}

sub showcreatetables {
  my $self = shift;;
  
  my $plugin = $self->stash->{plugin};
  my $command = $plugin->command;

  # Validation
  my $params = $command->params($self);
  my $rule = [
    database => {default => ''} => [
      'safety_name'
    ]
  ];
  my $vresult = $plugin->validator->validate($params, $rule);
  my $database = $vresult->data->{database};
  my $tables = $command->show_tables($database);
  
  # Get create tables
  my $create_tables = {};
  for my $table (@$tables) {
    $create_tables->{$table} = $plugin->command->show_create_table($database, $table);
  }
  
  $self->render(
    database => $database,
    create_tables => $create_tables
  );
}

sub showprimarykeys {
  my $self = shift;;
  
  my $plugin = $self->stash->{plugin};
  my $command = $plugin->command;

  # Validation
  my $params = $command->params($self);
  my $rule = [
    database => {default => ''} => [
      'safety_name'
    ],
  ];
  my $vresult = $plugin->validator->validate($params, $rule);
  my $database = $vresult->data->{database};
  
  # Get primary keys
  my $primary_keys = $command->show_primary_keys($database);
  
  $self->render(
    database => $database,
    primary_keys => $primary_keys
  );
}

sub shownullallowedcolumns {
  my $self = shift;;
  
  my $plugin = $self->stash->{plugin};
  my $command = $plugin->command;

  # Validation
  my $params = $command->params($self);
  my $rule = [
    database => {default => ''} => [
      'safety_name'
    ],
  ];
  my $vresult = $plugin->validator->validate($params, $rule);
  my $database = $vresult->data->{database};
  
  # Get null allowed columns
  my $null_allowed_columns = $command->show_null_allowed_columns($database);
  
  $self->render(
    database => $database,
    null_allowed_columns => $null_allowed_columns
  );
}

sub select {
  my $self = shift;;
  
  my $plugin = $self->stash->{plugin};
  my $command = $plugin->command;

  # Validation
  my $params = $command->params($self);
  my $rule = [
    database => {default => ''} => [
      'safety_name'
    ],
    table => {default => ''} => [
      'safety_name'
    ],
    page => {default => 1} => [
      'uint'
    ],
    condition_column => [
      'safety_name'
    ],
    condition_value => [
      'not_blank'
    ]
  ];
  my $vresult = $plugin->validator->validate($params, $rule);
  my $database = $vresult->data->{database};
  my $table = $vresult->data->{table};
  
  # Where
  my $column = $vresult->data->{condition_column};
  my $value = $vresult->data->{condition_value};
  
  my $where;
  if (defined $column && defined $value) {
    $where = $plugin->dbi->where;
    $where->clause(":${column}{like}");
    $where->param({$column => $value});
  }
  
  # Limit
  my $page = $vresult->data->{page};
  my $count = 100;
  my $offset = ($page - 1) * $count;
  
  # Get null allowed columns
  my $result = $plugin->dbi->select(
    table => "$database.$table",
    where => $where,
    append => "limit $offset, $count"
  );
  my $header = $result->header;
  my $rows = $result->fetch_all;
  my $sql = $plugin->dbi->last_sql;
  
  # Pager
  my $total = $plugin->dbi->select(
    'count(*)',
    table => "$database.$table",
    where => $where
  )->value;
  my $pager = Data::Page->new($total, $count, $page);
  
  $self->render(
    database => $database,
    table => $table,
    header => $header,
    rows => $rows,
    sql => $sql,
    pager => $pager
  );
}

1;


lib/Mojolicious/Plugin/MySQLViewerLite.pm
use 5.001001;
package Mojolicious::Plugin::MySQLViewerLite;
use Mojo::Base 'Mojolicious::Plugin::MySQLViewerLite::Base';
use File::Basename 'dirname';
use Cwd 'abs_path';
use Mojolicious::Plugin::MySQLViewerLite::Command;

our $VERSION = '0.15';

has command => sub {
  my $self = shift;
  my $commond = Mojolicious::Plugin::MySQLViewerLite::Command->new(dbi => $self->dbi);
};

sub register {
  my ($self, $app, $conf) = @_;
  my $prefix = $conf->{prefix} // 'mysqlviewerlite';
  
  # Database
  my $dbi = $conf->{dbi};
  my $connector = $conf->{connector};
  my $dbh = $conf->{dbh};
  if ($dbi) { $self->dbi($dbi) }
  elsif ($connector) { $self->dbi->connector($connector) }
  else { $self->dbi->dbh($dbh) }
  
  # Add template path
  $self->add_template_path($app->renderer, __PACKAGE__);
  
  # Routes
  my $r = $conf->{route} // $app->routes;
  $self->prefix($prefix);
  {
    my $r = $r->route("/$prefix")->to(
      'mysqlviewerlite#',
      namespace => 'Mojolicious::Plugin::MySQLViewerLite',
      plugin => $self,
      prefix => $self->prefix,
      main_title => 'MySQL Viewer Lite',
    );
    
    $r->get('/')->to('#default');
    $r->get('/tables')->to(
      '#tables',
      utilities => [
        {path => 'showcreatetables', title => 'Show create tables'},
        {path => 'showselecttables', title => 'Show select tables'},
        {path => 'showprimarykeys', title => 'Show primary keys'},
        {path => 'shownullallowedcolumns', title => 'Show null allowed columns'},
        {path => 'showdatabaseengines', title => 'Show database engines'},
        {path => 'showcharsets', title => 'Show charsets'}
      ]
    );
    $r->get('/table')->to('#table');
    $r->get('/showcreatetables')->to('#showcreatetables');
    $r->get('/showselecttables')->to('#showselecttables');
    $r->get('/showprimarykeys')->to('#showprimarykeys');
    $r->get('/shownullallowedcolumns')->to('#shownullallowedcolumns');
    $r->get('/showdatabaseengines')->to('#showdatabaseengines');
    $r->get('/showcharsets')->to('#showcharsets');
    $r->get('/select')->to('#select');

    # Routes (MySQL specific)
    $r->get('/showdatabaseengines')->to('#showdatabaseengines');
    $r->get('/showcharsets')->to('#showcharsets');
  }
}

1;

=head1 NAME

Mojolicious::Plugin::MySQLViewerLite - Mojolicious plugin to display MySQL database information on browser

=head1 CAUTION

B<L<Mojolicious::Plugin::MySQLViewerLite> is merged into L<Mojolicious::Plugin::DBViewer>>.

B<This module is DEPRECATED and will be removed from CPAN in 2018/4/1>.

But you get it on github.

  https://github.com/yuki-kimoto/Mojolicious-Plugin-MySQLViewerLite

=head1 SYNOPSYS

  # Mojolicious::Lite
  # (dbh is a database handle already connected to the database)
  plugin 'MySQLViewerLite', dbh => $dbh;

  # Mojolicious
  $app->plugin('MySQLViewerLite', dbh => $dbh);

  # Access
  http://localhost:3000/mysqlviewerlite
  
  # Prefix
  plugin 'MySQLViewerLite', dbh => $dbh, prefix => 'mysqlviewerlite2';

  # Route
  my $bridge = $app->route->under(sub {...});
  plugin 'MySQLViewerLite', dbh => $dbh, route => $bridge;

  # Using connection manager object instead of "dbh"
  plugin 'MySQLViewerLite', connector => DBIx::Connector->connect(...);

  # Using DBIx::Custom object instead of "dbh"
  plugin 'MySQLViewerLite', dbi => DBIx::Custom->connect(...);

=head1 DESCRIPTION

L<Mojolicious::Plugin::MySQLViewerLite> is L<Mojolicious> plugin
to display MySQL database information on your browser.

L<Mojolicious::Plugin::MySQLViewerLite> have the following features.

=over 4

=item *

Display all table names

=item *

Display C<show create table>

=item *

Select * from TABLE

=item *

Display C<primary keys>, C<null allowed columnes>, C<database engines> and C<charsets> in all tables.

=back

=head1 OPTIONS

=head2 C<connector>

  connector => $connector

Connector object such as L<DBIx::Connector> to connect to database.
You can use this instead of C<dbh> option.

  my $connector = DBIx::Connector->connect(...);

Connector has C<dbh> method to get database handle

=head2 C<dbh>

  dbh => $dbh

dbh is a L<DBI> database handle already connected to the database.

  my $dbh = DBI->connect(...);

=head2 C<dbi>

  dbi => DBIx::Custom->connect(...);

L<DBIx::Custom> object.
you can use this instead of C<dbh> option.

=head2 C<prefix>

  prefix => 'mysqlviewerlite2'

Application base path, default to C<mysqlviewerlite>.

=head2 C<route>

    route => $route

Router, default to C<$app->routes>.

It is useful when C<under> is used.

  my $bridge = $r->under(sub {...});
  plugin 'MySQLViewerLite', dbh => $dbh, route => $bridge;

=cut
